<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Differential Expression Analysis of Clonal Alterations Local effects â€¢ decal</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="apple-touch-icon-60x60.png">
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="bootstrap-toc.css">
<script src="bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="pkgdown.css" rel="stylesheet">
<script src="pkgdown.js"></script><meta property="og:title" content="Differential Expression Analysis of Clonal Alterations Local effects">
<meta property="og:description" content="It models gene UMI counts from a single-cell experiment as a
    negative binomial regression to measure the effect of a localized genetic
    alteration of a clonal cell subpopulation to near by genes expression.">
<meta property="og:image" content="https://mauranolab.github.io/decal/logo.png">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-home">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="index.html">decal</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.1.1</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="index.html">
    <span class="fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="articles/quick-start.html">Analyzing single-cell RNA-seq perturbation with DECAL</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="contents col-md-9">


<div id="decal-" class="section level1">
<div class="page-header"><h1 class="hasAnchor">
<a href="#decal-" class="anchor"></a>DECAL <img src="reference/figures/logo.png" align="right" width="150px">
</h1></div>
<p><strong>DECAL</strong> (<strong>D</strong>ifferential <strong>E</strong>xpression analysis of <strong>C</strong>lonal <strong>A</strong>lterations <strong>L</strong>ocal effects) provide you tools to conduct differential expression analysis of single-cell perturbations to potential interacting genes. Similar to other expression analysis tools, it models gene expression using a <em>Negative Binomial</em> (or <em>Gamma-Poisson</em>) regression, modeling each gene UMI count by the cell total count and the cell alteration status.</p>
<div id="features" class="section level2">
<h2 class="hasAnchor">
<a href="#features" class="anchor"></a>Features</h2>
<ul>
<li>
<code>decal</code> is compatible with <strong>tidyverse</strong> analysis.</li>
<li>Includes a suit of simulation functions to generate your own dataset based <code>decal</code> model and evaluate your statistical power.</li>
<li>Package has few dependencies, requiring only <code>MASS</code>, <code>fastglm</code> and <code>Matrix</code>.</li>
<li>Can evaluate specific clonal alteration and gene effect, instead of modeling all genes and all alterations. This allow us to quickly investigate a large number of interactions skipping unlikely effects.</li>
</ul>
</div>
<div id="installation" class="section level2">
<h2 class="hasAnchor">
<a href="#installation" class="anchor"></a>Installation</h2>
<p>To install <code>decal</code> package current version, open your R terminal and type:</p>
<pre class="{r}"><code>## Install remotes if not available with
## install.install.packages("remotes")
remotes::install_github("mauranolab/decal")</code></pre>
</div>
<div id="using-decal" class="section level2">
<h2 class="hasAnchor">
<a href="#using-decal" class="anchor"></a>Using <code>decal</code>
</h2>
<p>The package main function (<code>decal</code>) runs the whole statistical analysis by fitting a <em>Negative Binomial</em> (or <em>Gamma-Poisson</em>) regression to each gene and alteration pair specified and evaluate the perturbation statistical significance for a single-cell perturbation experiment. You can run <code>decal</code> as follow:</p>
<pre class="{r}"><code>decal(perturbations, count, clone)</code></pre>
<p>The three parameters required are the following:</p>
<ul>
<li>a table specifying the population and gene to evaluate perturbation (<code>perturbations</code>).</li>
<li>a UMI count matrix that each column correspond to a cell and each row a gene or feature (<code>count</code>).</li>
<li>a list of cells specifying your clones composition (<code>clone</code>). This is represented by a R list composed by character (or integer) vectors named after the clone ID and indicating the cells belonging to this clone.</li>
</ul>
<p><code>decal</code> returns a deep copy of the <code>perturbations</code> table with the following additional columns:</p>
<ul>
<li>
<code>n0</code> and <code>n1</code>: number of non-perturbed and perturbed cells, respectively.</li>
<li>
<code>x0</code> and <code>x1</code>: average UMI count among perturbed and non-perturbed cells, respectively.</li>
<li>
<code>mu</code>: average UMI count among all cells.</li>
<li>
<code>xb</code>: expected average UMI count of perturbed cells.</li>
<li>
<code>theta</code>: estimated <em>negative binomial</em> dispersion parameter.</li>
<li>
<code>z</code>: estimated perturbation z-score.</li>
<li>
<code>lfc</code>: <em>log2 fold-change</em> of perturbed cells gene expression.</li>
<li>
<code>pvalue</code> and <code>p_adjusted</code>: perturbation <em>t-test</em> significance values.</li>
</ul>
</div>
<div id="quick-start" class="section level2">
<h2 class="hasAnchor">
<a href="#quick-start" class="anchor"></a>Quick Start</h2>
<p>Below we include a example of <code>decal</code> analysis on a simulated dataset included with our package.</p>
<pre class="{r}"><code><a href="https://mauranolab.github.io/decal">library(decal)

data("sim_decal")
perturbations &lt;- sim_decal$perturbations
count &lt;- sim_decal$count
clone &lt;- sim_decal$clone

decal(perturbations, count, clone)</a></code></pre>
<p>Further details exploring this example are included in our <em>quick-start</em> vignette.</p>
</div>
<div id="acknowledgement" class="section level2">
<h2 class="hasAnchor">
<a href="#acknowledgement" class="anchor"></a>Acknowledgement</h2>
<ul>
<li>The negative binomial dispersion estimation process was extracted and modified from <a href="https://github.com/ChristophH/sctransform"><code>sctransform</code></a> package and described by <em>Hafemeister &amp; Satija (2019)</em>
</li>
</ul>
</div>
</div>

  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <div class="license">
<h2>License</h2>
<ul class="list-unstyled">
<li><a href="LICENSE.html">Full license</a></li>
<li><small><a href="https://opensource.org/licenses/mit-license.php">MIT</a> + file <a href="LICENSE-text.html">LICENSE</a></small></li>
</ul>
</div>
<div class="developers">
<h2>Developers</h2>
<ul class="list-unstyled">
<li>Andre Ribeiro-dos-Santos <br><small class="roles"> Author, maintainer </small> <a href="https://orcid.org/https://orcid.org/0000-0002-7005-3576" target="orcid.widget" aria-label="ORCID"><span class="fab fa-orcid orcid" aria-hidden="true"></span></a> </li>
</ul>
</div>

  <div class="dev-status">
<h2>Dev status</h2>
<ul class="list-unstyled">
<li><a href="https://github.com/mauranolab/decal/actions"><img src="https://github.com/mauranolab/decal/workflows/R-CMD-check/badge.svg" alt="R-CMD-check"></a></li>
<li><a href="https://CRAN.R-project.org/package=decal"><img src="https://www.r-pkg.org/badges/version/decal" alt="CRAN status"></a></li>
<li><a href="https://lifecycle.r-lib.org/articles/stages.html#experimental"><img src="https://img.shields.io/badge/lifecycle-experimental-orange.svg" alt="Lifecycle: experimental"></a></li>
</ul>
</div>
</div>
</div>


      <footer><div class="copyright">
  <p>Developed by Andre Ribeiro-dos-Santos.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
