<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Analyzing single-cell RNA-seq perturbation with DECAL • decal</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png">
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Analyzing single-cell RNA-seq perturbation with DECAL">
<meta property="og:description" content="decal">
<meta property="og:image" content="https://mauranolab.github.io/decal/logo.png">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">decal</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.1.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/quick-start.html">Analyzing single-cell RNA-seq perturbation with DECAL</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><script src="quick-start_files/accessible-code-block-0.0.1/empty-anchor.js"></script><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Analyzing single-cell RNA-seq perturbation with DECAL</h1>
                        <h4 class="author">André M. Ribeiro-dos-Santos</h4>
            
            <h4 class="date">07/26/2021</h4>
      
      
      <div class="hidden name"><code>quick-start.Rmd</code></div>

    </div>

    
        <div class="abstract">
      <p class="abstract">Abstract</p>
      <p>TODO: Package abstract decal package version: 0.1.0</p>
    </div>
    
<div id="default-workflow" class="section level1">
<h1 class="hasAnchor">
<a href="#default-workflow" class="anchor"></a>Default Workflow</h1>
<p><strong>DECAL</strong> (<strong>D</strong>ifferential <strong>E</strong>xpression analysis of <strong>C</strong>lonal <strong>A</strong>lterations <strong>L</strong>ocal effects) provide you with tools to conduct differential expression analysis of single-cell perturbations experiments to potentially interacting genes. Similar to other differential expression analysis tools, it models gene expression using a <em>Negative Binomial</em> (or <em>Gamma-Poisson</em>) regression, modeling each gene UMI count by the cell alteration status and cell total count.</p>
<div id="features" class="section level2">
<h2 class="hasAnchor">
<a href="#features" class="anchor"></a>Features</h2>
<ul>
<li>
<code>decal</code> is compatible with <strong>tidyverse</strong> analysis.</li>
<li>Includes a suit of simulation functions to generate your own dataset based <code>decal</code> model and evaluate your statistical power.</li>
<li>Package has few dependencies, requiring only <code>MASS</code>, <code>fastglm</code> and <code>Matrix</code>.</li>
<li>Can evaluate specific clonal alteration and gene effect, instead of modeling all genes and all alterations. This allow us to quickly investigate a large number of interactions skipping unlikely effects.</li>
</ul>
</div>
<div id="installation" class="section level2">
<h2 class="hasAnchor">
<a href="#installation" class="anchor"></a>Installation</h2>
<p>To install <code>decal</code> package current version, open your R terminal and type:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">## Install remotes if not available with</span>
<span class="co">## install.install.packages("remotes")</span>
<span class="fu">remotes</span><span class="fu">::</span><span class="fu"><a href="https://remotes.r-lib.org/reference/install_github.html">install_github</a></span><span class="op">(</span><span class="st">"mauranolab/decal"</span><span class="op">)</span></code></pre></div>
</div>
<div id="using-decal" class="section level2">
<h2 class="hasAnchor">
<a href="#using-decal" class="anchor"></a>Using <code>decal</code>
</h2>
<p>The package main function (<code>decal</code>) runs the whole statistical analysis by fitting a <em>Negative Binomial</em> (or <em>Gamma-Poisson</em>) regression to each gene and alteration pair specified and evaluate the perturbation statistical significance for a single-cell perturbation experiment. You can run <code>decal</code> as follow:</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/decal.html">decal</a></span><span class="op">(</span><span class="va">perturbations</span>, <span class="va">count</span>, <span class="va">clone</span><span class="op">)</span></code></pre></div>
<p>The three parameters required are the following:</p>
<ul>
<li>a table specifying the population and gene to evaluate perturbation (<code>perturbations</code>).</li>
<li>a UMI count matrix that each column correspond to a cell and each row a gene or feature (<code>count</code>).</li>
<li>a list of cells specifying your clones composition (<code>clone</code>). This is represented by a R list composed by character (or integer) vectors named after the clone ID and indicating the cells belonging to this clone.</li>
</ul>
<p><code>decal</code> returns a deep copy of the <code>perturbations</code> table with the following additional columns:</p>
<ul>
<li>
<code>n0</code> and <code>n1</code>: number of non-perturbed and perturbed cells, respectively.</li>
<li>
<code>x0</code> and <code>x1</code>: average UMI count among perturbed and non-perturbed cells, respectively.</li>
<li>
<code>mu</code>: average UMI count among all cells.</li>
<li>
<code>xb</code>: expected average UMI count of perturbed cells.</li>
<li>
<code>theta</code>: estimated <em>negative binomial</em> dispersion parameter.</li>
<li>
<code>z</code>: estimated perturbation z-score.</li>
<li>
<code>lfc</code>: <em>log2 fold-change</em> of perturbed cells gene expression.</li>
<li>
<code>pvalue</code> and <code>p_adjusted</code>: perturbation <em>t-test</em> significance values.</li>
</ul>
</div>
<div id="quick-start" class="section level2">
<h2 class="hasAnchor">
<a href="#quick-start" class="anchor"></a>Quick Start</h2>
<p>Below we include a example of <code>decal</code> analysis on a simulated dataset included with our package.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://mauranolab.github.io/decal">decal</a></span><span class="op">)</span>

<span class="fu"><a href="https://rdrr.io/r/utils/data.html">data</a></span><span class="op">(</span><span class="st">"sim_decal"</span><span class="op">)</span>
<span class="va">perturbations</span> <span class="op">&lt;-</span> <span class="va">sim_decal</span><span class="op">$</span><span class="va">perturbations</span>
<span class="va">count</span> <span class="op">&lt;-</span> <span class="va">sim_decal</span><span class="op">$</span><span class="va">count</span>
<span class="va">clone</span> <span class="op">&lt;-</span> <span class="va">sim_decal</span><span class="op">$</span><span class="va">clone</span>

<span class="va">res</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/decal.html">decal</a></span><span class="op">(</span><span class="va">perturbations</span>, <span class="va">count</span>, <span class="va">clone</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">res</span><span class="op">)</span>
<span class="co">#&gt;       clone     gene expected_lfc  n0 n1   x0   x1   mu theta   xb      z</span>
<span class="co">#&gt; 1 clone-001 gene1353            0 979 21 57.3 60.0 57.4    78 60.0  1.161</span>
<span class="co">#&gt; 2 clone-001 gene1212            0 979 21 37.3 38.0 37.3    73 38.5  0.631</span>
<span class="co">#&gt; 3 clone-001 gene1032            0 979 21 21.6 22.0 21.6    76 22.2  0.462</span>
<span class="co">#&gt; 4 clone-001  gene504            0 979 21  4.3  4.2  4.3    46  4.3 -0.044</span>
<span class="co">#&gt; 5 clone-001 gene1065            0 979 21 23.9 22.8 23.9    75 22.7 -0.959</span>
<span class="co">#&gt; 6 clone-001  gene995            0 979 21 19.4 19.7 19.4    77 19.7  0.270</span>
<span class="co">#&gt;       lfc pvalue p_adjusted</span>
<span class="co">#&gt; 1  0.0619   0.25       0.38</span>
<span class="co">#&gt; 2  0.0412   0.53       0.69</span>
<span class="co">#&gt; 3  0.0364   0.64       0.78</span>
<span class="co">#&gt; 4 -0.0071   0.96       0.98</span>
<span class="co">#&gt; 5 -0.0734   0.34       0.49</span>
<span class="co">#&gt; 6  0.0208   0.79       0.87</span></code></pre></div>
</div>
<div id="exploring-decal-results" class="section level2">
<h2 class="hasAnchor">
<a href="#exploring-decal-results" class="anchor"></a>Exploring decal results</h2>
<p>Now let’s explore <code>decal</code> results. Admitting a 5% <em>false discovery rate</em> (<em>FDR</em>), we can select significantly differentiate perturbations by picking those with <code>p_adjusted &lt; 0.05</code>. Below we present the top 5 interactions with highest reducing and increasing log fold-change effect (as measured by <code>lfc</code>).</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">sig</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/subset.html">subset</a></span><span class="op">(</span><span class="va">res</span>, <span class="va">p_adjusted</span> <span class="op">&lt;</span> <span class="fl">0.05</span><span class="op">)</span>
<span class="va">sig</span> <span class="op">&lt;-</span> <span class="va">sig</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/order.html">order</a></span><span class="op">(</span><span class="va">sig</span><span class="op">$</span><span class="va">lfc</span><span class="op">)</span>, <span class="op">]</span>
<span class="co">## select first 5 and last 5 since`sig` is ordered by lfc</span>
<span class="va">pick</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fl">5</span>, <span class="fu"><a href="https://rdrr.io/r/base/nrow.html">nrow</a></span><span class="op">(</span><span class="va">sig</span><span class="op">)</span> <span class="op">-</span> <span class="fl">1</span><span class="op">:</span><span class="fl">5</span><span class="op">)</span>
<span class="va">sig</span><span class="op">[</span><span class="va">pick</span>,<span class="op">]</span>
<span class="co">#&gt;         clone     gene expected_lfc  n0 n1   x0      x1   mu theta      xb    z</span>
<span class="co">#&gt; 352 clone-008  gene130           -3 972 28  1.4   0.036  1.4    18   0.033 -3.8</span>
<span class="co">#&gt; 83  clone-002  gene094           -2 986 14  1.2   0.071  1.2    17   0.082 -2.7</span>
<span class="co">#&gt; 572 clone-013  gene265           -3 970 30  2.0   0.133  1.9    22   0.134 -5.7</span>
<span class="co">#&gt; 483 clone-011  gene993           -3 995  5 19.8   1.200 19.7    77   1.414 -6.3</span>
<span class="co">#&gt; 88  clone-002  gene790           -3 986 14 10.1   0.714 10.0    68   0.823 -8.7</span>
<span class="co">#&gt; 471 clone-011 gene1207            3 995  5 37.4 264.400 38.5    73 314.135 20.0</span>
<span class="co">#&gt; 648 clone-015 gene1069            3 979 21 24.0 194.238 27.6    73 199.427 57.6</span>
<span class="co">#&gt; 120 clone-003  gene699            3 981 19  7.5  69.474  8.7    67  62.670 40.3</span>
<span class="co">#&gt; 604 clone-014  gene636            3 992  8  6.4  52.625  6.8    61  52.833 23.8</span>
<span class="co">#&gt; 76  clone-002  gene150            3 986 14  1.5  10.500  1.6    19  12.073 19.7</span>
<span class="co">#&gt;      lfc   pvalue p_adjusted</span>
<span class="co">#&gt; 352 -5.4  1.7e-04    3.7e-04</span>
<span class="co">#&gt; 83  -3.9  6.9e-03    1.4e-02</span>
<span class="co">#&gt; 572 -3.9  1.2e-08    3.1e-08</span>
<span class="co">#&gt; 483 -3.8  3.4e-10    9.2e-10</span>
<span class="co">#&gt; 88  -3.6  1.5e-17    4.6e-17</span>
<span class="co">#&gt; 471  3.1  6.9e-75    5.0e-74</span>
<span class="co">#&gt; 648  3.1 7.1e-320   5.7e-318</span>
<span class="co">#&gt; 120  3.1 1.5e-211   4.4e-210</span>
<span class="co">#&gt; 604  3.0 7.8e-100    7.2e-99</span>
<span class="co">#&gt; 76   3.0  4.2e-73    2.9e-72</span></code></pre></div>
<p>To visualize the perturbation effect, let’s plot the top 3 interactions increasing and reducing gene expression.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">## compute a normalize size-factor</span>
<span class="va">sf</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/mean.html">mean</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/colSums.html">colSums</a></span><span class="op">(</span><span class="va">count</span><span class="op">)</span><span class="op">)</span> <span class="op">/</span> <span class="fu"><a href="https://rdrr.io/r/base/colSums.html">colSums</a></span><span class="op">(</span><span class="va">count</span><span class="op">)</span>
<span class="co">## add log10(x+1) transformation function</span>
<span class="va">log10p</span> <span class="op">&lt;-</span> <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="fu"><a href="https://rdrr.io/r/base/Log.html">log10</a></span><span class="op">(</span><span class="va">x</span> <span class="op">+</span> <span class="fl">1</span><span class="op">)</span>

<span class="co">## pick tables first 3 and last 3</span>
<span class="va">pick</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fl">3</span>, <span class="fu"><a href="https://rdrr.io/r/base/nrow.html">nrow</a></span><span class="op">(</span><span class="va">sig</span><span class="op">)</span> <span class="op">-</span> <span class="fl">1</span><span class="op">:</span><span class="fl">3</span><span class="op">)</span>

<span class="fu"><a href="https://rdrr.io/r/graphics/par.html">par</a></span><span class="op">(</span>mfrow <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">2</span>, <span class="fl">3</span><span class="op">)</span><span class="op">)</span>
<span class="kw">for</span> <span class="op">(</span><span class="va">i</span> <span class="kw">in</span> <span class="va">pick</span><span class="op">)</span> <span class="op">{</span>
  <span class="co">## Determinate perturbed and unperturbed cells</span>
  <span class="va">x</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span><span class="op">(</span><span class="va">count</span><span class="op">)</span> <span class="op">%in%</span> <span class="va">clone</span><span class="op">[[</span><span class="va">sig</span><span class="op">$</span><span class="va">clone</span><span class="op">[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span><span class="op">]</span>
  <span class="co">## Calculate normalized expression count</span>
  <span class="va">y</span> <span class="op">&lt;-</span> <span class="va">count</span><span class="op">[</span><span class="va">sig</span><span class="op">$</span><span class="va">gene</span><span class="op">[</span><span class="va">i</span><span class="op">]</span>,<span class="op">]</span> <span class="op">*</span> <span class="va">sf</span>
  <span class="co">## plot jitter</span>
  <span class="va">title</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span><span class="op">(</span><span class="va">sig</span><span class="op">$</span><span class="va">clone</span><span class="op">[</span><span class="va">i</span><span class="op">]</span>, <span class="st">"+"</span>, <span class="va">sig</span><span class="op">$</span><span class="va">gene</span><span class="op">[</span><span class="va">i</span><span class="op">]</span>,<span class="st">"\nLFC: "</span>, <span class="fu"><a href="https://rdrr.io/r/base/Round.html">round</a></span><span class="op">(</span><span class="va">sig</span><span class="op">$</span><span class="va">lfc</span><span class="op">[</span><span class="va">i</span><span class="op">]</span>, <span class="fl">2</span><span class="op">)</span><span class="op">)</span>
  <span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html">plot</a></span><span class="op">(</span>
    <span class="va">x</span> <span class="op">+</span> <span class="fu"><a href="https://rdrr.io/r/stats/Uniform.html">runif</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/length.html">length</a></span><span class="op">(</span><span class="va">x</span><span class="op">)</span>, <span class="op">-</span><span class="fl">.3</span>, <span class="fl">.3</span><span class="op">)</span>, <span class="fu">log10p</span><span class="op">(</span><span class="va">y</span><span class="op">)</span>, main <span class="op">=</span> <span class="va">title</span>,
    axes <span class="op">=</span> <span class="cn">FALSE</span>, xlab <span class="op">=</span> <span class="st">""</span>, ylab <span class="op">=</span> <span class="st">"Normalize UMI count"</span>,
    xlim <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="op">-</span><span class="fl">.5</span>, <span class="fl">1.5</span><span class="op">)</span>, ylim <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">0.9</span>, <span class="fl">1.1</span><span class="op">)</span> <span class="op">*</span> <span class="fu"><a href="https://rdrr.io/r/base/range.html">range</a></span><span class="op">(</span><span class="fu">log10p</span><span class="op">(</span><span class="va">y</span><span class="op">)</span><span class="op">)</span>
  <span class="op">)</span>
  <span class="fu"><a href="https://rdrr.io/r/graphics/axis.html">axis</a></span><span class="op">(</span><span class="fl">1</span>, at <span class="op">=</span> <span class="fl">0</span><span class="op">:</span><span class="fl">1</span>, labels <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"perturbed"</span>, <span class="st">"unperturbed"</span><span class="op">)</span>, las <span class="op">=</span> <span class="fl">2</span><span class="op">)</span>
  <span class="fu"><a href="https://rdrr.io/r/graphics/axis.html">axis</a></span><span class="op">(</span><span class="fl">2</span>, at <span class="op">=</span> <span class="fu">log10p</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">2</span>, <span class="fl">5</span>, <span class="fl">10</span>, <span class="fl">20</span>, <span class="fl">50</span>, <span class="fl">100</span><span class="op">)</span><span class="op">)</span>, labels <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">2</span>, <span class="fl">5</span>, <span class="fl">10</span>, <span class="fl">20</span>, <span class="fl">50</span>, <span class="fl">100</span><span class="op">)</span><span class="op">)</span>
  <span class="co">## add 95% error bar</span>
  <span class="va">ymed</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="va">sig</span><span class="op">$</span><span class="va">mu</span><span class="op">[</span><span class="va">i</span><span class="op">]</span>, <span class="va">sig</span><span class="op">$</span><span class="va">xb</span><span class="op">[</span><span class="va">i</span><span class="op">]</span><span class="op">)</span>
  <span class="va">ylow</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/NegBinomial.html">qnbinom</a></span><span class="op">(</span><span class="fl">0.025</span>, size <span class="op">=</span> <span class="va">sig</span><span class="op">$</span><span class="va">theta</span><span class="op">[</span><span class="va">i</span><span class="op">]</span>, mu <span class="op">=</span> <span class="va">ymed</span><span class="op">)</span>
  <span class="va">yupr</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/NegBinomial.html">qnbinom</a></span><span class="op">(</span><span class="fl">0.975</span>, size <span class="op">=</span> <span class="va">sig</span><span class="op">$</span><span class="va">theta</span><span class="op">[</span><span class="va">i</span><span class="op">]</span>, mu <span class="op">=</span> <span class="va">ymed</span><span class="op">)</span>
  <span class="fu"><a href="https://rdrr.io/r/graphics/points.html">points</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">1</span><span class="op">)</span>, <span class="fu">log10p</span><span class="op">(</span><span class="va">ymed</span><span class="op">)</span>, col <span class="op">=</span> <span class="st">"firebrick"</span>, cex <span class="op">=</span> <span class="fl">2</span>, pch <span class="op">=</span> <span class="fl">16</span><span class="op">)</span>
  <span class="fu"><a href="https://rdrr.io/r/graphics/arrows.html">arrows</a></span><span class="op">(</span>
    x0 <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">1</span><span class="op">)</span>, y0 <span class="op">=</span> <span class="fu">log10p</span><span class="op">(</span><span class="va">ylow</span><span class="op">)</span>, x1 <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">1</span><span class="op">)</span>, y1 <span class="op">=</span> <span class="fu">log10p</span><span class="op">(</span><span class="va">yupr</span><span class="op">)</span>,
    code <span class="op">=</span> <span class="fl">3</span>, angle <span class="op">=</span> <span class="fl">90</span>, col <span class="op">=</span> <span class="st">"firebrick"</span>, length <span class="op">=</span> <span class="fl">.1</span>, lwd <span class="op">=</span> <span class="fl">2</span>
  <span class="op">)</span>
<span class="op">}</span></code></pre></div>
<p><img src="quick-start_files/figure-html/quick_start_plot-1.png" width="700"></p>
</div>
<div id="evaluating-decal-results" class="section level2">
<h2 class="hasAnchor">
<a href="#evaluating-decal-results" class="anchor"></a>Evaluating decal results</h2>
<p>In our <code>sim_decal</code> simulated dataset, we introduced some real expression perturbations as indicated by <code>expected_lfc</code> column in <code>perturbations</code> table, such as a 0 <code>expected_lfc</code> indicate no change was applied. We can evaluate our results using the confusion matrix below:</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">conf_mat</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/table.html">table</a></span><span class="op">(</span>real <span class="op">=</span> <span class="va">res</span><span class="op">$</span><span class="va">expected_lfc</span> <span class="op">!=</span> <span class="fl">0</span>, decal <span class="op">=</span> <span class="va">res</span><span class="op">$</span><span class="va">p_adjusted</span> <span class="op">&lt;</span> <span class="fl">0.05</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/Round.html">round</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/proportions.html">prop.table</a></span><span class="op">(</span><span class="va">conf_mat</span><span class="op">)</span>, <span class="fl">3</span><span class="op">)</span>
<span class="co">#&gt;        decal</span>
<span class="co">#&gt; real    FALSE  TRUE</span>
<span class="co">#&gt;   FALSE 0.447 0.009</span>
<span class="co">#&gt;   TRUE  0.026 0.518</span></code></pre></div>
<p>Based on this confusion matrix, our <code>decal</code> model presented:</p>
<ul>
<li>
<strong>Accuracy</strong>: 0.96</li>
<li>
<strong>Precision</strong>: 0.95</li>
<li>
<strong>Recall</strong>: 0.98</li>
<li>
<strong>F1 Score</strong>: 0.97</li>
<li>
<strong>Observed FDR</strong>: 0.04</li>
</ul>
<p>These results show great metrics with recall, accuracy, precision and F1 score above 95% and a <em>false discovery ratio</em> within our admitted 5%.</p>
<p>Next, let’s explore how well <code>decal</code> estimated the real <em>log2 fold-change</em> (LFC) applied. Illustrated below is the decal estimated LFC (<code>lfc</code> column in <code>res</code>) distribution for each of the real perturbations values applied (<code>expected_lfc</code> column).</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">lfc_histogram</span> <span class="op">&lt;-</span> <span class="kw">function</span><span class="op">(</span><span class="va">lfc</span>, <span class="va">fn</span> <span class="op">=</span> <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="op">{</span> <span class="va">x</span><span class="op">$</span><span class="va">count</span> <span class="op">/</span> <span class="fu"><a href="https://rdrr.io/r/base/sum.html">sum</a></span><span class="op">(</span><span class="va">x</span><span class="op">$</span><span class="va">count</span><span class="op">)</span> <span class="op">}</span><span class="op">)</span> <span class="op">{</span>
  <span class="va">hist_breaks</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="op">-</span><span class="cn">Inf</span>, <span class="fu"><a href="https://rdrr.io/r/base/seq.html">seq</a></span><span class="op">(</span><span class="op">-</span><span class="fl">4</span>, <span class="fl">4</span>, length.out <span class="op">=</span> <span class="fl">50</span><span class="op">)</span>, <span class="cn">Inf</span><span class="op">)</span>
  <span class="va">hist_counts</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html">lapply</a></span><span class="op">(</span><span class="va">lfc</span>, <span class="va">hist</span>, breaks <span class="op">=</span> <span class="va">hist_breaks</span>, plot <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span>
  <span class="va">hist_ymax</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/Extremes.html">max</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/lapply.html">sapply</a></span><span class="op">(</span><span class="va">hist_counts</span>, <span class="va">fn</span><span class="op">)</span><span class="op">)</span>
  <span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html">plot</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="op">-</span><span class="fl">4</span>, <span class="fl">4</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">0</span>, <span class="va">hist_ymax</span><span class="op">)</span>, type <span class="op">=</span> <span class="st">"n"</span>,
    xlab <span class="op">=</span> <span class="st">"Estimated LFC"</span>, ylab <span class="op">=</span> <span class="st">"Proportion of interactions"</span><span class="op">)</span>
  <span class="kw">for</span><span class="op">(</span><span class="va">i</span> <span class="kw">in</span> <span class="fu"><a href="https://rdrr.io/r/base/seq.html">seq_along</a></span><span class="op">(</span><span class="va">hist_counts</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span>
    <span class="fu"><a href="https://rdrr.io/r/graphics/lines.html">lines</a></span><span class="op">(</span><span class="va">hist_counts</span><span class="op">[[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span><span class="op">$</span><span class="va">mids</span>, <span class="fu">fn</span><span class="op">(</span><span class="va">hist_counts</span><span class="op">[[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span><span class="op">)</span>, col <span class="op">=</span> <span class="va">i</span>, type <span class="op">=</span> <span class="st">"s"</span><span class="op">)</span>
  <span class="op">}</span>
  <span class="fu"><a href="https://rdrr.io/r/graphics/legend.html">legend</a></span><span class="op">(</span><span class="st">"topleft"</span>, title <span class="op">=</span> <span class="st">"Real LFC"</span>, legend <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">hist_counts</span><span class="op">)</span>,
    col <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/seq.html">seq_along</a></span><span class="op">(</span><span class="va">hist_counts</span><span class="op">)</span>, bty <span class="op">=</span> <span class="st">"n"</span>, lty <span class="op">=</span> <span class="fl">1</span>, lwd <span class="op">=</span> <span class="fl">2</span>, ncol <span class="op">=</span> <span class="fl">2</span><span class="op">)</span>
<span class="op">}</span>

<span class="fu">lfc_histogram</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/split.html">split</a></span><span class="op">(</span><span class="va">res</span><span class="op">$</span><span class="va">lfc</span>, <span class="va">res</span><span class="op">$</span><span class="va">expected_lfc</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<p><img src="quick-start_files/figure-html/quick_start_hist-1.png" width="700"></p>
<p>Finally, we can measure decal LFC estimative <em>Mean Squared Error</em> (<em>MSE</em>) and <em>Rooted MSE</em> (<em>RMSE</em>) using the code below. The results indicate that <code>decal</code> estimates presented high accuracy and precision with <em>RMSE</em> below 1 LFC for all perturbations effects applied.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">err</span> <span class="op">&lt;-</span> <span class="op">(</span><span class="va">res</span><span class="op">$</span><span class="va">lfc</span> <span class="op">-</span> <span class="va">res</span><span class="op">$</span><span class="va">expected_lfc</span><span class="op">)</span><span class="op">**</span><span class="fl">2</span>
<span class="va">err</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/split.html">split</a></span><span class="op">(</span><span class="va">err</span>, <span class="va">res</span><span class="op">$</span><span class="va">expected_lfc</span><span class="op">)</span>
<span class="va">mse</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html">sapply</a></span><span class="op">(</span><span class="va">err</span>, <span class="va">mean</span>, na.rm <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>

<span class="fu"><a href="https://rdrr.io/r/base/data.frame.html">data.frame</a></span><span class="op">(</span>
  N <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html">sapply</a></span><span class="op">(</span><span class="va">err</span>, <span class="va">length</span><span class="op">)</span>,
  MSE <span class="op">=</span> <span class="va">mse</span>, RMSE <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/MathFun.html">sqrt</a></span><span class="op">(</span><span class="va">mse</span><span class="op">)</span>
<span class="op">)</span>
<span class="co">#&gt;      N    MSE  RMSE</span>
<span class="co">#&gt; -3  30 0.3118 0.558</span>
<span class="co">#&gt; -2  75 0.1852 0.430</span>
<span class="co">#&gt; -1  75 0.0758 0.275</span>
<span class="co">#&gt; 0  300 0.0422 0.205</span>
<span class="co">#&gt; 1   75 0.0171 0.131</span>
<span class="co">#&gt; 2   75 0.0120 0.109</span>
<span class="co">#&gt; 3   30 0.0029 0.054</span></code></pre></div>
</div>
</div>
<div id="exploring-decal-arguments" class="section level1">
<h1 class="hasAnchor">
<a href="#exploring-decal-arguments" class="anchor"></a>Exploring <code>decal</code> arguments</h1>
<p>As mentioned before, <code>decal</code> function requires three main parameters:</p>
<ol style="list-style-type: decimal">
<li>
<code>perturbation</code>, a table specifying the clone population and gene to evaluate differential expression.</li>
<li>
<code>count</code>, a UMI count matrix where each column correspond to a cell and each row a gene or feature.</li>
<li>
<code>clone</code>, a list of cells specifying each clone populations composition.</li>
</ol>
<p>In the next sections, we will further illustrate and give alternative formats for (1) and (3).</p>
<div id="electing-potential-gene-perturbations" class="section level2">
<h2 class="hasAnchor">
<a href="#electing-potential-gene-perturbations" class="anchor"></a>Electing potential gene perturbations</h2>
<p>Since <code>decal</code> was designed to explore specific gene perturbations, it requires you to specify clone and gene (or features) pairs to evaluate expression change among the cells within the clone in regards to all other cells. For example, in our original study after we localized our alternations and evaluated their impact to all genes within <em>250 kbp</em>.</p>
<p>In <code>sim_decal$perturbations</code> data.frame (see snippet below) the clone and gene pairs are specified by the <code>clone</code> and <code>gene</code> columns, which can be an id (as character vector) or index (as integer vector). In the case these elements are specified as an index, they identify the ith element of <code>sim_decal$clone</code> and row position in <code>sim_decal$count</code>, respectively.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">sim_decal</span><span class="op">$</span><span class="va">perturbations</span><span class="op">)</span>
<span class="co">#&gt;       clone     gene expected_lfc</span>
<span class="co">#&gt; 1 clone-001 gene1353            0</span>
<span class="co">#&gt; 2 clone-001 gene1212            0</span>
<span class="co">#&gt; 3 clone-001 gene1032            0</span>
<span class="co">#&gt; 4 clone-001  gene504            0</span>
<span class="co">#&gt; 5 clone-001 gene1065            0</span>
<span class="co">#&gt; 6 clone-001  gene995            0</span></code></pre></div>
<p>By default <code>decal</code> uses the <code>clone</code> and <code>gene</code> columns, but you can specify a different column using <code>gene_col</code> and <code>clone_col</code> parameters like in the example below.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">alt_pert</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html">data.frame</a></span><span class="op">(</span>gene_ix <span class="op">=</span> <span class="fl">101</span><span class="op">:</span><span class="fl">110</span>, clone_ix <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">10</span><span class="op">)</span>
<span class="fu"><a href="../reference/decal.html">decal</a></span><span class="op">(</span><span class="va">alt_pert</span>, <span class="va">count</span>, <span class="va">clone</span>,
  gene_col <span class="op">=</span> <span class="st">"gene_ix"</span>, clone_col <span class="op">=</span> <span class="st">"clone_ix"</span><span class="op">)</span>
<span class="co">#&gt;           gene_ix clone_ix  n0 n1  x0   x1  mu theta   xb      z     lfc pvalue</span>
<span class="co">#&gt; clone-001     101        1 979 21 1.3 1.05 1.3    17 1.06 -0.777 -0.2441  0.438</span>
<span class="co">#&gt; clone-002     102        2 986 14 1.2 1.00 1.2    17 1.15 -0.294 -0.1098  0.769</span>
<span class="co">#&gt; clone-003     103        3 981 19 1.2 1.42 1.2    17 1.22 -0.063 -0.0185  0.950</span>
<span class="co">#&gt; clone-004     104        4 981 19 1.2 1.11 1.2    17 1.04 -0.811 -0.2542  0.417</span>
<span class="co">#&gt; clone-005     105        5 976 24 1.2 1.58 1.2    17 1.58  1.500  0.3733  0.134</span>
<span class="co">#&gt; clone-006     106        6 993  7 1.2 1.00 1.2    17 1.37  0.278  0.1488  0.781</span>
<span class="co">#&gt; clone-007     107        7 989 11 1.3 0.73 1.3    18 0.81 -1.296 -0.6740  0.195</span>
<span class="co">#&gt; clone-008     108        8 972 28 1.2 1.79 1.3    17 1.63  1.835  0.3912  0.067</span>
<span class="co">#&gt; clone-009     109        9 979 21 1.3 1.38 1.3    18 1.34  0.028  0.0081  0.978</span>
<span class="co">#&gt; clone-010     110       10 975 25 1.3 1.40 1.3    18 1.26 -0.210 -0.0525  0.834</span>
<span class="co">#&gt;           p_adjusted</span>
<span class="co">#&gt; clone-001       0.88</span>
<span class="co">#&gt; clone-002       0.98</span>
<span class="co">#&gt; clone-003       0.98</span>
<span class="co">#&gt; clone-004       0.88</span>
<span class="co">#&gt; clone-005       0.65</span>
<span class="co">#&gt; clone-006       0.98</span>
<span class="co">#&gt; clone-007       0.65</span>
<span class="co">#&gt; clone-008       0.65</span>
<span class="co">#&gt; clone-009       0.98</span>
<span class="co">#&gt; clone-010       0.98</span></code></pre></div>
</div>
<div id="defining-your-experiment-clonal-structure" class="section level2">
<h2 class="hasAnchor">
<a href="#defining-your-experiment-clonal-structure" class="anchor"></a>Defining your experiment clonal structure</h2>
<p>After identifying your cells’ clonal populations you can specify it <code>decal</code> (<code>clone</code> parameter) as a R list where each element is clone represented by a collection of ids (as character vector) or index (as integer vector) indicating the set of cells belonging to the clone. When the cells are specified by index, they correspond to the column position in your <code>count</code> matrix. See below a example from <code>sim_decal$clone</code></p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">sim_decal</span><span class="op">$</span><span class="va">clone</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">2</span><span class="op">]</span>
<span class="co">#&gt; $`clone-001`</span>
<span class="co">#&gt;  [1] "cell998" "cell741" "cell558" "cell339" "cell085" "cell701" "cell096"</span>
<span class="co">#&gt;  [8] "cell843" "cell477" "cell410" "cell675" "cell379" "cell323" "cell620"</span>
<span class="co">#&gt; [15] "cell232" "cell213" "cell830" "cell676" "cell117" "cell784" "cell352"</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; $`clone-002`</span>
<span class="co">#&gt;  [1] "cell438" "cell443" "cell464" "cell420" "cell922" "cell424" "cell704"</span>
<span class="co">#&gt;  [8] "cell149" "cell665" "cell214" "cell569" "cell444" "cell315" "cell955"</span></code></pre></div>
</div>
<div id="reducing-noise-and-increasing-power" class="section level2">
<h2 class="hasAnchor">
<a href="#reducing-noise-and-increasing-power" class="anchor"></a>Reducing noise and increasing power</h2>
<p>To improve your findings, <code>decal</code> includes some filtering arguments to avoid testing interactions with low statistical power:</p>
<ul>
<li>
<code>min_x</code>: minimal average UMI count on perturbed and unperturbed cells (indicated by <code>x1</code> and <code>x0</code> on resulting table, respectively).</li>
<li>
<code>min_n</code>: minimal number of perturbed cells (indicated by <code>n1</code>).</li>
<li>
<code>min_mu</code>: minimal global average count (indicated by <code>mu</code>), also skips <code>theta</code> estimation.</li>
</ul>
<p>Interactions that doesn’t met all of these requirements are skipped and no differential expression analysis is conducted, eventhough they are no removed from the results. By default, <code>decal</code> applies the following minimal filters <code>min_x = 1</code>; <code>min_n = 2</code>; and <code>min_mu = 0.05</code>.</p>
</div>
</div>
<div id="understanding-decal-model" class="section level1">
<h1 class="hasAnchor">
<a href="#understanding-decal-model" class="anchor"></a>Understanding <code>decal</code> model</h1>
<div id="statistical-model" class="section level2">
<h2 class="hasAnchor">
<a href="#statistical-model" class="anchor"></a>Statistical model</h2>
<p>The <code>decal</code> statistical model is based on the observations that UMI counts of a particular gene approximates a <em>Poisson</em> or <em>Negative Binomial</em> distribution in a single-cell RNA-seq experiment <span class="citation">(Svensson 2020; Townes et al. 2019)</span>. Thus, we proposed to model the observed counts as follow:</p>
<p><span class="math display">\[
Y_{gc} \sim NB(\mu_{gci}, \theta_g)
(\#eq:model)
\]</span></p>
<p>where <span class="math inline">\(Y_{cg}\)</span> is the UMI count for gene <em>g</em> and cell <em>c</em> modeled using a <em>negative binomial</em> distribution with expected count <span class="math inline">\(\mu_{cgi}\)</span> and a gene-specific dispersion parameter <span class="math inline">\(\theta_g\)</span>.</p>
<p><span class="math display">\[
log(\mu_{gci}) = log(D_c) + \beta_g + \beta_x X_{ci}
(\#eq:modelMu)
\]</span></p>
<p>The <span class="math inline">\(log(\mu_{cgi})\)</span> (or log expected count) is proportional to the perturbation effect (<span class="math inline">\(\beta_x\)</span>) of a specific clone <em>i</em> indicated by <span class="math inline">\(X_{ci}\)</span> (where <span class="math inline">\(X_{ci} == 1\)</span> if cell <em>c</em> belongs to the perturbed clone or <span class="math inline">\(X_{ci} == 0\)</span> otherwise) and offseted by the log cell total UMI count (<span class="math inline">\(D_c = \sum_g Y_{cg}\)</span>).</p>
<p>The parameter <span class="math inline">\(\theta_g\)</span> defines the model dispersion where smaller values produce a wider distribution and higher values produce a tighter distribution that coincides with a <em>Poisson</em> distribution. Such as the model variance is defined as follow:</p>
<p><span class="math display">\[
\begin{aligned}
Var(Y_{cg}) &amp; = E[(Y_{cg} - \mu_{cgi})^2] \\
 &amp; = \mu_{cgi} + \mu_{cgi}^2/\theta_g \\
\lim_{\theta\to\infty} NB(\mu, \theta) &amp; = Pois(\mu)
\end{aligned}
(\#eq:variance)
\]</span></p>
<p>This modeling approach is similar to previously published models such as edgeR <span class="citation">(Robinson, McCarthy, and Smyth 2010; McCarthy, Chen, and Smyth 2012)</span>, DESeq2 <span class="citation">(Anders and Huber 2010; Love, Huber, and Anders 2014)</span>, and glmGamPoi <span class="citation">(Ahlmann-Eltze and Huber 2021)</span>. But it differs in two points: (i) to estimate <span class="math inline">\(\theta_g\)</span> we adopted a regularized estimation strategy to make it robust to sampling noise in low expression genes; and (ii) instead of fitting the model to all genes, it fits and measure specific perturbation to gene.</p>
</div>
<div id="estimating-dispersion" class="section level2">
<h2 class="hasAnchor">
<a href="#estimating-dispersion" class="anchor"></a>Estimating dispersion</h2>
<p>To estimate and regularize our <span class="math inline">\(\theta_g\)</span>, we adopted the strategy described by <span class="citation">Hafemeister and Satija (2019)</span>. This strategy consists in first selecting about 2000 genes, estimating their expect count (<span class="math inline">\(mu_{cg}\)</span>) using a naive Poisson regression offseted by the log cell total count, such as:</p>
<p><span class="math display">\[
\begin{aligned}
Y_{cg} &amp; \sim Pois(mu_{cg}) \\
log(mu_{cg}) &amp; = \beta_g + log(D_c)
\end{aligned}
(#eq:naiveTheta)
\]</span></p>
<p>Next, we use <span class="math inline">\(mu_{cg}\)</span> to make a naive estimate of <span class="math inline">\(\theta_g\)</span> using a maximum likelihood estimator (<code>MASS::theta_ml</code> function). The resulting naive estimate was finally regularized by fitting it to a kernel smooth regression as a function of the average gene count (<span class="math inline">\(\sum_g Y_{cg} / N\)</span>) to produce our final estimate of <span class="math inline">\(\theta_g\)</span>. Illustrate below is a scatterplot of each naive estimate, in red our final regularized estimate and in black the real dispersion applied when simulating <code>sim_decal</code> dataset.</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">raw_theta</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/attr.html">attr</a></span><span class="op">(</span><span class="va">res</span>, <span class="st">"raw_theta"</span><span class="op">)</span>
<span class="va">mu</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colSums.html">rowMeans</a></span><span class="op">(</span><span class="va">count</span><span class="op">)</span>
<span class="va">mu_breaks</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">0.01</span>, <span class="fl">0.1</span>, <span class="fl">1</span>, <span class="fl">2</span>, <span class="fl">5</span>, <span class="fl">10</span>, <span class="fl">20</span>, <span class="fl">50</span>, <span class="fl">100</span><span class="op">)</span>

<span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html">plot</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/Log.html">log10</a></span><span class="op">(</span><span class="va">mu</span><span class="op">)</span>, <span class="va">raw_theta</span>,
  xlab <span class="op">=</span> <span class="st">"Average expression (mu)"</span>, xaxt <span class="op">=</span> <span class="st">"n"</span>,
  ylab <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/expression.html">expression</a></span><span class="op">(</span><span class="va">theta</span><span class="op">)</span>, ylim <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">150</span><span class="op">)</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/graphics/abline.html">abline</a></span><span class="op">(</span>h <span class="op">=</span> <span class="fl">100</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/graphics/lines.html">lines</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/Log.html">log10</a></span><span class="op">(</span><span class="va">res</span><span class="op">$</span><span class="va">mu</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/order.html">order</a></span><span class="op">(</span><span class="va">res</span><span class="op">$</span><span class="va">mu</span><span class="op">)</span><span class="op">]</span><span class="op">)</span>, <span class="va">res</span><span class="op">$</span><span class="va">theta</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/order.html">order</a></span><span class="op">(</span><span class="va">res</span><span class="op">$</span><span class="va">mu</span><span class="op">)</span><span class="op">]</span>, col <span class="op">=</span> <span class="fl">2</span>, lwd <span class="op">=</span> <span class="fl">3</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/graphics/axis.html">axis</a></span><span class="op">(</span><span class="fl">1</span>, at <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/Log.html">log10</a></span><span class="op">(</span><span class="va">mu_breaks</span><span class="op">)</span>, labels <span class="op">=</span> <span class="va">mu_breaks</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/graphics/legend.html">legend</a></span><span class="op">(</span>
  <span class="st">"topleft"</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"Naive Estimate"</span>, <span class="st">"Regularized Estimate"</span>, <span class="st">"Real"</span><span class="op">)</span>,
  pch <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="cn">NA</span>, <span class="cn">NA</span><span class="op">)</span>, lty <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="cn">NA</span>, <span class="fl">1</span>, <span class="fl">1</span><span class="op">)</span>, col <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">2</span>, <span class="fl">1</span><span class="op">)</span>
<span class="op">)</span></code></pre></div>
<p><img src="quick-start_files/figure-html/unnamed-chunk-2-1.png" width="700"></p>
<p>As shown above, our regularized estimates approximates the real dispersion values and control some of the excessive fluctuation presented by our naive estimate.</p>
</div>
<div id="assumptions-and-caveats" class="section level2">
<h2 class="hasAnchor">
<a href="#assumptions-and-caveats" class="anchor"></a>Assumptions and Caveats</h2>
<p>It is worth notice that <code>decal</code> statiscal model implies some assumptions and caveats:</p>
<ol style="list-style-type: decimal">
<li>The UMI count distribution is independent between genes, which is usually true for single-cell experiments due to large quantity of genes and small proportion of counts each of them represent.</li>
<li>The effect of individual clones are not enough to affect the unperturbed estimates for other clones. For such assumption to hold, the number of perturbated cells of any clone must be much smaller than the total number of cells available (<code>n1 &lt;&lt; N</code>). This allow us to explore each clone independently, since it makes the effect of other clones to our unperturbed estimates negletable.</li>
<li>The cells investigates must represent an uniform population differing only by the induced alterations. This can safely be assumed when the experiment is conducted in a uniform cell culture.</li>
</ol>
</div>
</div>
<div id="simulating" class="section level1">
<h1 class="hasAnchor">
<a href="#simulating" class="anchor"></a>Simulating</h1>
<p>Since many factor can affect your experiment, we included to <code>decal</code> a way to simulate a dataset under our model assumptions (see equation @ref(eq:model)). These tools will allow you to estimate your experiment <strong>power</strong> and <strong>sensitivity</strong> under different conditions you may find.</p>
<p><code>decal</code> simulation suite is compose of various functions to generate a count matrix (<code>sim_count</code> and <code>sim_count_from_data</code>), randomly assign cells to different clones (<code>sim_clone</code> and <code>sim_clone_range</code>), and generate a full experiment dataset (<code>sim_experiment</code> and <code>sim_experiment_from_data</code>). Here we will use <code>sim_experiment</code> and <code>sim_experiment_from_data</code> to generate our examples, but the other functions behave similarly.</p>
<div id="evaluating-our-model-under-null-hypothesis" class="section level2">
<h2 class="hasAnchor">
<a href="#evaluating-our-model-under-null-hypothesis" class="anchor"></a>Evaluating our model under null hypothesis</h2>
<p>Under the null hypothesis, we expect that our perturbations produce no expression change to nearby genes. Admitting a <em>FDR</em> of 5% and that <code>decal</code> statistical model holds, we would expect that the resulting z-score present a normal distribution centered at 0, p-value have a uniform distribution and about 5% of the tests below our significance threshold.</p>
<p>Given an existing dataset, we can evaluate if our model holds by randomly recombining clone and gene pairs.</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">rnd</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/unique.html">unique</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/data.frame.html">data.frame</a></span><span class="op">(</span>
  clone <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/sample.html">sample</a></span><span class="op">(</span><span class="va">perturbations</span><span class="op">$</span><span class="va">clone</span><span class="op">)</span>,
  gene <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/sample.html">sample</a></span><span class="op">(</span><span class="va">perturbations</span><span class="op">$</span><span class="va">gene</span><span class="op">)</span>
<span class="op">)</span><span class="op">)</span>
<span class="co">## remove previous clone + gene pairs</span>
<span class="va">rnd</span> <span class="op">&lt;-</span> <span class="va">rnd</span><span class="op">[</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste</a></span><span class="op">(</span><span class="va">rnd</span><span class="op">$</span><span class="va">gene</span>, <span class="va">rnd</span><span class="op">$</span><span class="va">clone</span><span class="op">)</span> <span class="op">%in%</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste</a></span><span class="op">(</span><span class="va">perturbations</span><span class="op">$</span><span class="va">gene</span>, <span class="va">perturbations</span><span class="op">$</span><span class="va">clone</span><span class="op">)</span>,<span class="op">]</span>
<span class="va">rnd_res</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/decal.html">decal</a></span><span class="op">(</span><span class="va">rnd</span>, <span class="va">count</span>, <span class="va">clone</span><span class="op">)</span>
<span class="va">rnd_res</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/subset.html">subset</a></span><span class="op">(</span><span class="va">rnd_res</span>, <span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html">is.na</a></span><span class="op">(</span><span class="va">rnd_res</span><span class="op">$</span><span class="va">pvalue</span><span class="op">)</span><span class="op">)</span>

<span class="fu"><a href="https://rdrr.io/r/graphics/par.html">par</a></span><span class="op">(</span>mfrow <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">3</span><span class="op">)</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/graphics/hist.html">hist</a></span><span class="op">(</span><span class="va">rnd_res</span><span class="op">$</span><span class="va">z</span>, xlab <span class="op">=</span> <span class="st">"Z-score"</span>, main <span class="op">=</span> <span class="st">""</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/graphics/hist.html">hist</a></span><span class="op">(</span><span class="va">rnd_res</span><span class="op">$</span><span class="va">pvalue</span>, xlab <span class="op">=</span> <span class="st">"Observed Pvalue"</span>, main <span class="op">=</span> <span class="st">""</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/graphics/hist.html">hist</a></span><span class="op">(</span><span class="va">rnd_res</span><span class="op">$</span><span class="va">p_adjusted</span>, xlab <span class="op">=</span> <span class="st">"Observed Qvalue"</span>, main <span class="op">=</span> <span class="st">""</span><span class="op">)</span></code></pre></div>
<p><img src="quick-start_files/figure-html/sim_randomization-1.png" width="768"></p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R">
<span class="fu"><a href="https://rdrr.io/r/base/mean.html">mean</a></span><span class="op">(</span><span class="va">rnd_res</span><span class="op">$</span><span class="va">pvalue</span> <span class="op">&lt;</span> <span class="fl">0.05</span><span class="op">)</span>
<span class="co">#&gt; [1] 0.044</span>
<span class="fu"><a href="https://rdrr.io/r/base/mean.html">mean</a></span><span class="op">(</span><span class="va">rnd_res</span><span class="op">$</span><span class="va">p_adjusted</span> <span class="op">&lt;</span> <span class="fl">0.05</span><span class="op">)</span>
<span class="co">#&gt; [1] 0</span></code></pre></div>
<p>As expected, this experiment z-score presented the expected <em>z-score</em> and <em>p-value</em> distribution and the ratio of tests with p-value below the significance threshold (0.04) within the expected 5% at random.</p>
<p>We can conduct a similar experiment producing a random count matrix based on your experiment counts. <code>sim_count_from_data</code> will simulate a UMI count matrix with the same dimensions as your experiment with cells total UMI count and gene average expression similar to your reference count matrix.</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">rnd_count</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/simulate_count.html">sim_count_from_data</a></span><span class="op">(</span><span class="va">count</span><span class="op">)</span>

<span class="va">rnd_res</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/decal.html">decal</a></span><span class="op">(</span><span class="va">perturbations</span>, <span class="va">rnd_count</span>, <span class="va">clone</span><span class="op">)</span>
<span class="va">rnd_res</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/subset.html">subset</a></span><span class="op">(</span><span class="va">rnd_res</span>, <span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html">is.na</a></span><span class="op">(</span><span class="va">rnd_res</span><span class="op">$</span><span class="va">pvalue</span><span class="op">)</span><span class="op">)</span>

<span class="fu"><a href="https://rdrr.io/r/graphics/par.html">par</a></span><span class="op">(</span>mfrow <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">3</span><span class="op">)</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/graphics/hist.html">hist</a></span><span class="op">(</span><span class="va">rnd_res</span><span class="op">$</span><span class="va">z</span>, xlab <span class="op">=</span> <span class="st">"Z-score"</span>, main <span class="op">=</span> <span class="st">""</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/graphics/hist.html">hist</a></span><span class="op">(</span><span class="va">rnd_res</span><span class="op">$</span><span class="va">pvalue</span>, xlab <span class="op">=</span> <span class="st">"Observed Pvalue"</span>, main <span class="op">=</span> <span class="st">""</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/graphics/hist.html">hist</a></span><span class="op">(</span><span class="va">rnd_res</span><span class="op">$</span><span class="va">p_adjusted</span>, xlab <span class="op">=</span> <span class="st">"Observed Qvalue"</span>, main <span class="op">=</span> <span class="st">""</span><span class="op">)</span></code></pre></div>
<p><img src="quick-start_files/figure-html/unnamed-chunk-3-1.png" width="768"></p>
</div>
<div id="estimating-your-experiment-power" class="section level2">
<h2 class="hasAnchor">
<a href="#estimating-your-experiment-power" class="anchor"></a>Estimating your experiment power</h2>
<p>Another usufull application of <code>decal</code> simulation suite is conducting a power analysis to decide how much should you sequence or to define your tests filters. In the example below, we generated an experiment using <code>sim_experiment_from_data</code> which generates a random count matrix, a clone assignment list, and a perturbation table. The generated count matrix is based on your experiment matrix with the same dimensions, similar cells total count and gene average expression. The clone list is composed of 100 clones with 5 to 20 cells each. Perturbation table indicates pairs of clone and genes and the log2 fold-change applied, where for each clone a gene was randomly to apply a effect indicated by <code>lfc</code>. In this case, for each clone, 180 genes were selected and listed in <code>pwr_dat$perturbations</code> and among them 20 were perturbed by a log2 fold-change of -2, -1, 1, and 2.</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">pwr_lfc</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/rep.html">rep</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="op">-</span><span class="fl">2</span>, <span class="op">-</span><span class="fl">1</span>, <span class="fl">1</span>, <span class="fl">2</span><span class="op">)</span>, each <span class="op">=</span> <span class="fl">20</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/rep.html">rep</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">100</span><span class="op">)</span><span class="op">)</span>
<span class="va">pwr_dat</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/simulate_experiment.html">sim_experiment_from_data</a></span><span class="op">(</span>
  <span class="va">count</span>, lfc <span class="op">=</span> <span class="va">pwr_lfc</span>, nclones <span class="op">=</span> <span class="fl">100</span>, min_n <span class="op">=</span> <span class="fl">5</span>, max_n <span class="op">=</span> <span class="fl">20</span>
<span class="op">)</span>
<span class="va">pwr_res</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/decal.html">decal</a></span><span class="op">(</span><span class="va">pwr_dat</span><span class="op">$</span><span class="va">perturbations</span>, <span class="va">pwr_dat</span><span class="op">$</span><span class="va">count</span>, <span class="va">pwr_dat</span><span class="op">$</span><span class="va">clone</span><span class="op">)</span>
<span class="va">pwr_res</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/subset.html">subset</a></span><span class="op">(</span><span class="va">pwr_res</span>, <span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html">is.na</a></span><span class="op">(</span><span class="va">pvalue</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<p>Adminitting a 5% FDR, we can estimated the ratio of the truly perturbated gene and clone pairs were detected by <code>decal</code> (left plot) and the clone size frequency (right plot). For all log2 fold-change applied here, <code>decal</code> obtained a power close to 80% that indicate que performance of the model.</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/graphics/par.html">par</a></span><span class="op">(</span>mfrow <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">2</span><span class="op">)</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/graphics/barplot.html">barplot</a></span><span class="op">(</span>
  <span class="fu"><a href="https://rdrr.io/r/base/lapply.html">sapply</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/split.html">split</a></span><span class="op">(</span>
    <span class="va">pwr_res</span><span class="op">$</span><span class="va">p_adjusted</span><span class="op">[</span><span class="va">pwr_res</span><span class="op">$</span><span class="va">expected_lfc</span> <span class="op">!=</span> <span class="fl">0</span><span class="op">]</span> <span class="op">&lt;</span> <span class="fl">0.05</span>,
    <span class="va">pwr_res</span><span class="op">$</span><span class="va">expected_lfc</span><span class="op">[</span><span class="va">pwr_res</span><span class="op">$</span><span class="va">expected_lfc</span> <span class="op">!=</span> <span class="fl">0</span><span class="op">]</span>
  <span class="op">)</span>, <span class="va">mean</span><span class="op">)</span>,
  xlab <span class="op">=</span> <span class="st">"LFC"</span>, ylab <span class="op">=</span> <span class="st">"Power"</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/graphics/abline.html">abline</a></span><span class="op">(</span>h <span class="op">=</span> <span class="fl">0.8</span>, lty <span class="op">=</span> <span class="fl">2</span>, col <span class="op">=</span> <span class="st">"firebrick"</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/graphics/hist.html">hist</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/lapply.html">sapply</a></span><span class="op">(</span><span class="va">pwr_dat</span><span class="op">$</span><span class="va">clone</span>, <span class="va">length</span><span class="op">)</span>, main <span class="op">=</span> <span class="st">""</span>, xlab <span class="op">=</span> <span class="st">"Number of perturbed cells"</span><span class="op">)</span></code></pre></div>
<p><img src="quick-start_files/figure-html/unnamed-chunk-4-1.png" width="700"></p>
</div>
</div>
<div id="references" class="section level1">
<h1 class="hasAnchor">
<a href="#references" class="anchor"></a>References</h1>
<div id="refs">
<div id="ref-ahlmann-eltze_glmgampoi_2021">
<p>Ahlmann-Eltze, Constantin, and Wolfgang Huber. 2021. “glmGamPoi: Fitting Gamma-Poisson Generalized Linear Models on Single Cell Count Data.” <em>Bioinformatics (Oxford, England)</em> 36 (24): 5701–2. <a href="https://doi.org/10.1093/bioinformatics/btaa1009">https://doi.org/10.1093/bioinformatics/btaa1009</a>.</p>
</div>
<div id="ref-anders_differential_2010">
<p>Anders, Simon, and Wolfgang Huber. 2010. “Differential Expression Analysis for Sequence Count Data.” <em>Genome Biology</em> 11 (10): R106. <a href="https://doi.org/10.1186/gb-2010-11-10-r106">https://doi.org/10.1186/gb-2010-11-10-r106</a>.</p>
</div>
<div id="ref-hafemeister_normalization_2019">
<p>Hafemeister, Christoph, and Rahul Satija. 2019. “Normalization and Variance Stabilization of Single-Cell RNA-Seq Data Using Regularized Negative Binomial Regression.” <em>Genome Biology</em> 20 (1): 296. <a href="https://doi.org/10.1186/s13059-019-1874-1">https://doi.org/10.1186/s13059-019-1874-1</a>.</p>
</div>
<div id="ref-love_moderated_2014">
<p>Love, Michael I., Wolfgang Huber, and Simon Anders. 2014. “Moderated Estimation of Fold Change and Dispersion for RNA-Seq Data with DESeq2.” <em>Genome Biology</em> 15 (12): 550. <a href="https://doi.org/10.1186/s13059-014-0550-8">https://doi.org/10.1186/s13059-014-0550-8</a>.</p>
</div>
<div id="ref-mccarthy_differential_2012">
<p>McCarthy, Davis J., Yunshun Chen, and Gordon K. Smyth. 2012. “Differential Expression Analysis of Multifactor RNA-Seq Experiments with Respect to Biological Variation.” <em>Nucleic Acids Research</em> 40 (10): 4288–97. <a href="https://doi.org/10.1093/nar/gks042">https://doi.org/10.1093/nar/gks042</a>.</p>
</div>
<div id="ref-robinson_edger_2010">
<p>Robinson, Mark D., Davis J. McCarthy, and Gordon K. Smyth. 2010. “edgeR: A Bioconductor Package for Differential Expression Analysis of Digital Gene Expression Data.” <em>Bioinformatics (Oxford, England)</em> 26 (1): 139–40. <a href="https://doi.org/10.1093/bioinformatics/btp616">https://doi.org/10.1093/bioinformatics/btp616</a>.</p>
</div>
<div id="ref-svensson_droplet_2020">
<p>Svensson, Valentine. 2020. “Droplet scRNA-Seq Is Not Zero-Inflated.” <em>Nature Biotechnology</em> 38 (2): 147–50. <a href="https://doi.org/10.1038/s41587-019-0379-5">https://doi.org/10.1038/s41587-019-0379-5</a>.</p>
</div>
<div id="ref-townes_feature_2019">
<p>Townes, F. William, Stephanie C. Hicks, Martin J. Aryee, and Rafael A. Irizarry. 2019. “Feature Selection and Dimension Reduction for Single-Cell RNA-Seq Based on a Multinomial Model.” <em>Genome Biology</em> 20 (1): 295. <a href="https://doi.org/10.1186/s13059-019-1861-6">https://doi.org/10.1186/s13059-019-1861-6</a>.</p>
</div>
</div>
</div>
<div id="session-info" class="section level1">
<h1 class="hasAnchor">
<a href="#session-info" class="anchor"></a>Session info</h1>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html">sessionInfo</a></span><span class="op">(</span><span class="op">)</span>
<span class="co">#&gt; R version 4.1.0 (2021-05-18)</span>
<span class="co">#&gt; Platform: x86_64-apple-darwin17.0 (64-bit)</span>
<span class="co">#&gt; Running under: macOS Catalina 10.15.7</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Matrix products: default</span>
<span class="co">#&gt; BLAS:   /Library/Frameworks/R.framework/Versions/4.1/Resources/lib/libRblas.dylib</span>
<span class="co">#&gt; LAPACK: /Library/Frameworks/R.framework/Versions/4.1/Resources/lib/libRlapack.dylib</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; locale:</span>
<span class="co">#&gt; [1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; attached base packages:</span>
<span class="co">#&gt; [1] stats     graphics  grDevices utils     datasets  methods   base     </span>
<span class="co">#&gt; </span>
<span class="co">#&gt; other attached packages:</span>
<span class="co">#&gt; [1] decal_0.1.0</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; loaded via a namespace (and not attached):</span>
<span class="co">#&gt;  [1] Rcpp_1.0.7        knitr_1.33        magrittr_2.0.1    MASS_7.3-54      </span>
<span class="co">#&gt;  [5] lattice_0.20-44   R6_2.5.0          ragg_1.1.3        rlang_0.4.11     </span>
<span class="co">#&gt;  [9] fastmap_1.1.0     highr_0.9         stringr_1.4.0     tools_4.1.0      </span>
<span class="co">#&gt; [13] grid_4.1.0        xfun_0.24         htmltools_0.5.1.1 systemfonts_1.0.2</span>
<span class="co">#&gt; [17] yaml_2.2.1        digest_0.6.27     rprojroot_2.0.2   fastglm_0.0.1    </span>
<span class="co">#&gt; [21] pkgdown_1.6.1     crayon_1.4.1      bookdown_0.22     textshaping_0.3.5</span>
<span class="co">#&gt; [25] Matrix_1.3-3      fs_1.5.0          memoise_2.0.0     cachem_1.0.5     </span>
<span class="co">#&gt; [29] evaluate_0.14     rmarkdown_2.9     stringi_1.7.3     compiler_4.1.0   </span>
<span class="co">#&gt; [33] desc_1.3.0</span></code></pre></div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p>Developed by Andre Ribeiro-dos-Santos.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
